validator/sawtooth_validator/execution/context_manager.py

```diff
@@ -421,48 +607,120 @@ def run(self):
+                     |
+                     |-> set (F) ----> set
+                     |
+    Address in base: |-> set (F)
+      Doesn't exist -|
+                     |-> get Future doesn't exit in context
+                     |
+                     |-> get ----> set (F)
+
+    """
+
+    def __init__(self, address, result=None, wait_for_tree=False):
         self.address = address
-        self._result = None
-        self._result_is_set = False
+        self._result = result
+        self._result_set_in_context = False
         self._condition = Condition()
         self._wait_for_tree = wait_for_tree
         self._tree_has_set = False
+        self._read_only = False
 
-    def done(self):
-        return self._result_is_set
+    def make_read_only(self):
+        with self._condition:
+            if self._wait_for_tree and not self._result_set_in_context:
+                self._condition.wait_for(
+                    lambda: self._tree_has_set or self._result_set_in_context)
+
+            self._read_only = True
+
+    def set_in_context(self):
+        with self._condition:
+            return self._result_set_in_context
 
     def result(self):
+        """Return the value at an address, optionally waiting until it is
+        set from the context_manager, or set based on the pre-fetch mechanism.
+
+        Returns:
+            (bytes): The opaque value for an address.
+        """
+
+        if self._read_only:
+            return self._result
         with self._condition:
-            if self._wait_for_tree:
-                self._condition.wait_for(lambda: self._result_is_set)
+            if self._wait_for_tree and not self._result_set_in_context:
+                self._condition.wait_for(
+                    lambda: self._tree_has_set or self._result_set_in_context)
             return self._result
 
     def set_result(self, result, from_tree=False):
-        """Set the addresses's value. If the _ContextFuture needs to be
-        resolved from a read from the MerkleTree first, wait to set the value
-        until _tree_has_set is True.
+        """Set the addresses's value unless the future has been declared
+        read only.
 
         Args:
             result (bytes): The value at an address.
-            from_tree (bool): Whether the value is coming from the MerkleTree,
-                or if it is set from prior state.
+            from_tree (bool): Whether the value is being set by a read from
+                the merkle tree.
 
         Returns:
             None
         """
+
+        if self._read_only:
+            return
+
         with self._condition:
-            if self._wait_for_tree and not from_tree:
-                self._condition.wait_for(lambda: self._tree_has_set)
-            self._result = result
-            self._result_is_set = True
-            if self._wait_for_tree and from_tree:
-                self._tree_has_set = True
+            if self._read_only:
```

Comment:
I think there should be a separate read-only `_ContextFuture` that gets created, rather than transforming a read-write instance to be read-only, which is why this is here.